pgwire text encoding fixes

madelynnblue opened the pull request:
Most commits here are bug fixes or test improvements. Arrays/tuples and intervals got significant changes to match postgres (but they weren't broken before), and are benchmarked below.

Array encoding delta:

```
name                                                     old time/op  new time/op  delta
Encodings/array['']/text-8                                511ns ± 8%   475ns ± 4%     ~     (p=0.095 n=5+5)
Encodings/array['\x0d53e338548082'::BYTEA]/text-8         846ns ±10%   604ns ± 9%  -28.57%  (p=0.008 n=5+5)
Encodings/array['test_with_spaces'::BYTEA]/text-8         883ns ± 3%   719ns ±10%  -18.66%  (p=0.008 n=5+5)
Encodings/array['name'::NAME]/text-8                      893ns ± 6%   599ns ±11%  -32.95%  (p=0.008 n=5+5)
Encodings/array[NULL]::text[]/text-8                      615ns ± 9%   523ns ± 2%  -14.99%  (p=0.008 n=5+5)
Encodings/array['']::text[]/text-8                        587ns ± 1%   553ns ± 9%     ~     (p=0.151 n=5+5)
Encodings/array['test']::text[]/text-8                    651ns ± 3%   579ns ± 8%  -10.97%  (p=0.008 n=5+5)
Encodings/array['test_with_spaces']::text[]/text-8        837ns ± 9%   688ns ± 7%  -17.79%  (p=0.008 n=5+5)
Encodings/array[e'\f']::text[]/text-8                     634ns ± 2%   550ns ± 8%  -13.27%  (p=0.008 n=5+5)
Encodings/array[e'\uFEFF']::text[]/text-8                1.18µs ± 4%  0.57µs ± 3%  -51.94%  (p=0.008 n=5+5)
Encodings/array[e'\u2603']::text[]/text-8                 799ns ± 3%   573ns ± 7%  -28.25%  (p=0.008 n=5+5)
Encodings/(1::int8,2::int8,3::int8,4::int8,null)/text-8   488ns ± 7%   619ns ± 3%  +26.99%  (p=0.008 n=5+5)
Encodings/('test_with_spaces'::BYTEA,null)/text-8         720ns ±27%   460ns ± 8%  -36.12%  (p=0.008 n=5+5)
Encodings/('test_with_spaces'::TEXT,null)/text-8          517ns ± 9%   437ns ±10%  -15.51%  (p=0.008 n=5+5)
Encodings/('name'::NAME,null)/text-8                      525ns ± 6%   325ns ± 7%  -38.02%  (p=0.008 n=5+5)
Encodings/('false'::JSONB,null)/text-8                   1.04µs ± 7%  0.63µs ± 6%  -39.00%  (p=0.008 n=5+5)
Encodings/('{"a":_[]}'::JSONB,null)/text-8               1.25µs ± 6%  0.75µs ±13%  -40.44%  (p=0.008 n=5+5)
```

Interval encoding delta:

```
name                                                    old time/op  new time/op  delta
Encodings/'10y10mon'::interval/text-8                    836ns ± 4%  1055ns ± 4%   +26.19%  (p=0.008 n=5+5)
Encodings/'1y1mon'::interval/text-8                      830ns ± 6%  1019ns ±15%   +22.78%  (p=0.008 n=5+5)
Encodings/'23:12:34'::interval/text-8                   1.00µs ± 5%  1.10µs ± 9%    +9.18%  (p=0.032 n=5+5)
Encodings/'-23:00:00'::interval/text-8                   791ns ± 7%  1016ns ± 5%   +28.56%  (p=0.008 n=5+5)
Encodings/'-10d'::interval/text-8                        798ns ±12%   958ns ± 3%   +20.03%  (p=0.008 n=5+5)
Encodings/'1ms'::interval/text-8                         765ns ± 3%  1418ns ± 6%   +85.36%  (p=0.008 n=5+5)
Encodings/'.2ms'::interval/text-8                        777ns ± 3%  1387ns ± 4%   +78.39%  (p=0.008 n=5+5)
Encodings/'-1mon1m'::interval/text-8                     898ns ± 5%  1386ns ± 9%   +54.29%  (p=0.008 n=5+5)
Encodings/'-1y1m'::interval/text-8                       900ns ± 7%  1380ns ± 1%   +53.34%  (p=0.008 n=5+5)
Encodings/'3y4mon5d6ms'::interval/text-8                1.13µs ± 2%  2.23µs ±13%   +97.24%  (p=0.008 n=5+5)
Encodings/'296537y20d15h30m7s'::interval/text-8         1.35µs ± 8%  1.77µs ± 8%   +31.03%  (p=0.008 n=5+5)
Encodings/'-2965y_-20d_-15h_-30m_-7s'::interval/text-8  1.39µs ± 6%  1.67µs ± 4%   +19.92%  (p=0.008 n=5+5)
Encodings/'00:00:00'::interval/text-8                    566ns ± 3%   583ns ± 3%      ~     (p=0.175 n=5+5)
```

madelynnblue committed:
pgwire: enable text encoding tests in generate-binary

These don't pass as of this commit and are thus skipped.

Release note: None

madelynnblue committed:
pgwire: remove time output from dates

Release note (bug fix): dates no longer have a time component in their
text encoding over the wire.

madelynnblue committed:
pgwire: use postgres instead of Go interval encoding for pgwire

These changes match postgres for all tested inputs.

Benchmarks are slower across the board which makes sense due to the
div/mod nanosecond section that is now not conditional.

name                                                    old time/op  new time/op  delta
Encodings/'10y10mon'::interval/text-8                    836ns ± 4%  1055ns ± 4%   +26.19%  (p=0.008 n=5+5)
Encodings/'1y1mon'::interval/text-8                      830ns ± 6%  1019ns ±15%   +22.78%  (p=0.008 n=5+5)
Encodings/'23:12:34'::interval/text-8                   1.00µs ± 5%  1.10µs ± 9%    +9.18%  (p=0.032 n=5+5)
Encodings/'-23:00:00'::interval/text-8                   791ns ± 7%  1016ns ± 5%   +28.56%  (p=0.008 n=5+5)
Encodings/'-10d'::interval/text-8                        798ns ±12%   958ns ± 3%   +20.03%  (p=0.008 n=5+5)
Encodings/'1ms'::interval/text-8                         765ns ± 3%  1418ns ± 6%   +85.36%  (p=0.008 n=5+5)
Encodings/'.2ms'::interval/text-8                        777ns ± 3%  1387ns ± 4%   +78.39%  (p=0.008 n=5+5)
Encodings/'-1mon1m'::interval/text-8                     898ns ± 5%  1386ns ± 9%   +54.29%  (p=0.008 n=5+5)
Encodings/'-1y1m'::interval/text-8                       900ns ± 7%  1380ns ± 1%   +53.34%  (p=0.008 n=5+5)
Encodings/'3y4mon5d6ms'::interval/text-8                1.13µs ± 2%  2.23µs ±13%   +97.24%  (p=0.008 n=5+5)
Encodings/'296537y20d15h30m7s'::interval/text-8         1.35µs ± 8%  1.77µs ± 8%   +31.03%  (p=0.008 n=5+5)
Encodings/'-2965y_-20d_-15h_-30m_-7s'::interval/text-8  1.39µs ± 6%  1.67µs ± 4%   +19.92%  (p=0.008 n=5+5)
Encodings/'00:00:00'::interval/text-8                    566ns ± 3%   583ns ± 3%      ~     (p=0.175 n=5+5)

Release note (bug fix): intervals now match postgres in their text
encoding over the wire.

Release note (bug fix): intervals no longer sometimes lose 1ns of
precision. This only happened rarely due to floating point inaccuracy.

madelynnblue committed:
tree: change TestEval to use a data directory

Using the giant list is cumbersome and makes fixing output
difficult. Using datadriven gives us some additional organization and
allows us to use the -rewrite flag when the format of a datum changes.

madelynnblue committed:
pgwire: test timezones and BC for timestamp

timestamptz differs slightly but isn't broken, which is why it's
commented out.

madelynnblue committed:
pgwire: fix bit encoding for 0 length

madelynnblue committed:
pgwire: fix array and tuple encoding

Match postgres encoding much more closely for tuples and arrays.

Implement binary tuple encoding. It's pretty simple: number of elements
then each element's OID and binary encoding.

Text encoding for arrays and tuples has some fairly simple quoting rules:
whitespace, quotes, or commas. We were also encoding unicode runes more
than needed.

Unwray datums to remove any OID oddities. This is already done in
non-array/tuple encoding.

Change the encoding tests to specify exactly the bytes the Text format
wants. We still keep the text format around in the encodings.json file
just for human consumption. Since text editors have problems displaying
weirdo things like the byte order mark it's safer to just print out
the expected bytes as numbers instead of hoping there's a value there
that our editors can't show. But in any case switch to %q for the Text
display to try to improve the debugability anyway.

The benchmark below is faster in cases where the type in the array
or tuple is a string. I think this is because of the removed call to
unicode.IsGraphic on each rune. The slower benchmark is mostly ints,
which now go through the same escape and quote check logic as strings,
which would explain the slowdown. We could special case those to improve
performance further.

name                                                     old time/op  new time/op  delta
Encodings/array['']/text-8                                511ns ± 8%   475ns ± 4%     ~     (p=0.095 n=5+5)
Encodings/array['\x0d53e338548082'::BYTEA]/text-8         846ns ±10%   604ns ± 9%  -28.57%  (p=0.008 n=5+5)
Encodings/array['test_with_spaces'::BYTEA]/text-8         883ns ± 3%   719ns ±10%  -18.66%  (p=0.008 n=5+5)
Encodings/array['name'::NAME]/text-8                      893ns ± 6%   599ns ±11%  -32.95%  (p=0.008 n=5+5)
Encodings/array[NULL]::text[]/text-8                      615ns ± 9%   523ns ± 2%  -14.99%  (p=0.008 n=5+5)
Encodings/array['']::text[]/text-8                        587ns ± 1%   553ns ± 9%     ~     (p=0.151 n=5+5)
Encodings/array['test']::text[]/text-8                    651ns ± 3%   579ns ± 8%  -10.97%  (p=0.008 n=5+5)
Encodings/array['test_with_spaces']::text[]/text-8        837ns ± 9%   688ns ± 7%  -17.79%  (p=0.008 n=5+5)
Encodings/array[e'\f']::text[]/text-8                     634ns ± 2%   550ns ± 8%  -13.27%  (p=0.008 n=5+5)
Encodings/array[e'\uFEFF']::text[]/text-8                1.18µs ± 4%  0.57µs ± 3%  -51.94%  (p=0.008 n=5+5)
Encodings/array[e'\u2603']::text[]/text-8                 799ns ± 3%   573ns ± 7%  -28.25%  (p=0.008 n=5+5)
Encodings/(1::int8,2::int8,3::int8,4::int8,null)/text-8   488ns ± 7%   619ns ± 3%  +26.99%  (p=0.008 n=5+5)
Encodings/('test_with_spaces'::BYTEA,null)/text-8         720ns ±27%   460ns ± 8%  -36.12%  (p=0.008 n=5+5)
Encodings/('test_with_spaces'::TEXT,null)/text-8          517ns ± 9%   437ns ±10%  -15.51%  (p=0.008 n=5+5)
Encodings/('name'::NAME,null)/text-8                      525ns ± 6%   325ns ± 7%  -38.02%  (p=0.008 n=5+5)
Encodings/('false'::JSONB,null)/text-8                   1.04µs ± 7%  0.63µs ± 6%  -39.00%  (p=0.008 n=5+5)
Encodings/('{"a":_[]}'::JSONB,null)/text-8               1.25µs ± 6%  0.75µs ±13%  -40.44%  (p=0.008 n=5+5)

Fixes #31847

Release note (bug fix): Correct pgwire encoding for arrays and tuples.

madelynnblue committed:
pgwire: verify encoding length instead of ignoring it in TestEncodings

Improved output and smarter type handling as well.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32144)
<!-- Reviewable:end -->

mjibson commented:
Ok PTAL when you have time, no rush. Also I'm going to be out until the 19th so feel free to do whatever you want including merging or waiting until I come back to finish this up.

mjibson commented:
Rebased onto master with the previous PR commits now included, so all commits here need review. Thanks.

BramGruneir reviewed (changes requested):
For the first commit.  It seems wrong to me to first commit a bunch of failing tests and then the fixes for them.  Could you reorder this so the tests go in after they should work.  Or add in a skip for now that you remove once the tests pass.
The general rule is that each commit should move the state of the system from one good state to a better one.  This breaks that by adding failing tests.  

I love the commit (the 3rd one) that moves this to a list of data files.  And I'm trusting that these contain the full list suite of tests from the original.  But I do think that could have been pull out into its own PR.

The 4th PR should be re-reviewed by @bobvawter, since he's now the expert on dates.  Also, does this change it so we have to specify dates and interval in the postgres way?  Or can we still also use the go way?  Would be a shame to not be able to do both. 

Got through all but the last commit so far.  Will continue tomorrow.

> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 37 of 37 files at r3, 36 of 36 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, 3 of 3 files at r7, 4 of 30 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-:-LR8uSrY6Sbj3Si7S1fm:b-8l0ebh)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cli/testdata/dump/row, line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8zm9K1gDyEhe_jP0x:-LR8zm9K1gDyEhe_jP0y:b-gq4h57) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/cli/testdata/dump/row#L93)):*
> ```
> 	(NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
> ----
> ----
> ```

Was the extra empty line purposefully removed?
___
*[pkg/sql/logictest/testdata/logic_test/time, line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR90ZkuBBrNUuB2sQZj:-LR90ZkuBBrNUuB2sQZk:b-v19w9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/sql/logictest/testdata/logic_test/time#L194)):*
> ```
> SELECT '11:59:59':::TIME - '12:00:00':::TIME
> ----
> -00:00:01
> ```

How do we handle `-00:00:00`?  Is it the same as `00:00:00`?  Please add tests for both of them.
___
*[pkg/sql/pgwire/types.go, line 197 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR91DtH02rvvZHhSFN0:-LR91DtH02rvvZHhSFN1:bgk0cis) ([raw file](https://github.com/cockroachdb/cockroach/blob/b052b7f6c93469861b9b2d9f63b07ccedb05b368/pkg/sql/pgwire/types.go#L197)):*
> ```Go
> 	case *tree.DBitArray:
> 		words, lastBitsUsed := v.EncodingParts()
> 		if len(words) == 0 {
> ```

Can you add a test for this?  Also, a comment here would be nice.
___
*[pkg/sql/sem/tree/testdata/eval/overflow, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8xAB34aiZ9nu1upAq:-LR8xAB34aiZ9nu1upAr:b-qc97m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026045098067c7d48d993ec293725ab4c79dacd/pkg/sql/sem/tree/testdata/eval/overflow#L1)):*
> ```
> eval
> ```

no first line with an explanation
___
*[pkg/sql/sem/tree/testdata/eval/row, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8xOiy6DUTxhwbVj9h:-LR8xOiy6DUTxhwbVj9i:b-n61mtj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026045098067c7d48d993ec293725ab4c79dacd/pkg/sql/sem/tree/testdata/eval/row#L1)):*
> ```
> # Row
> ```

most of these have a space after the first comment
___
*[pkg/sql/sem/tree/testdata/eval/unary, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8xTFn15TXWW6J5sSb:-LR8xTFn15TXWW6J5sSc:b-42ylyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026045098067c7d48d993ec293725ab4c79dacd/pkg/sql/sem/tree/testdata/eval/unary#L1)):*
> ```
> eval
> ```

No first line comment
___
*[pkg/util/duration/duration.go, line 174 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8zIlVEcZqk0cPaT-e:-LR8zIlVEcZqk0cPaT-f:bnpfy8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/util/duration/duration.go#L174)):*
> ```Go
> 
> 	wrote := false
> 	wrotePrev := func() {
> ```

I like this way of dealing with adding spaces.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Overall LGTM on the changes, modulo resolution of the open comments.

However I agree with Bram that the sequence of commits should be amenable to a git bisect, where the test suite is always expected to succeed.

Also I get that the change in the last comment is useful/important, but the extra work performed is quite the bummer. Could you please include a before/after benchmark of the encoding throughput for arrays/tuples on that commit.

> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 37 of 37 files at r3, 36 of 36 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, 3 of 3 files at r7, 29 of 30 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-:-LRSX_wA5496qq7YKDI2:b4jqvp8)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/cmp-protocol/main.go, line 68 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSWuy4CFNMvZln614F:-LRSWuy4CFNMvZln614G:b-8iimce) ([raw file](https://github.com/cockroachdb/cockroach/blob/77ce7deda4c688ccc50f1a2ecee0f074baf06cd6/pkg/cmd/cmp-protocol/main.go#L68)):*
> ```Go
> 					sqlbase.ColumnType_OIDVECTOR,
> 					sqlbase.ColumnType_OID,         // our 8-byte ints are usually out of range for pg
> 					sqlbase.ColumnType_FLOAT,       // slight rounding differences at the end
> ```

1) Is there a way to include them, but have some kind of function which masks the last digits during the comparison to assert whether the test successful?


2) I'd like to see tests for DECIMAL and FLOAT which exercise the encodings for infinities and NaN. Are there any already? Perhaps somewhere else?
___
*[pkg/sql/sem/tree/interval.go, line 562 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSUzvp2dB1vVy65Jot:-LRSUzvp2dB1vVy65Jou:b-iqrrb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/sql/sem/tree/interval.go#L562)):*
> ```Go
> // or ".123" to "123000000". This function takes care to round correctly
> // when a naive conversion would incorrectly truncate due to floating point
> // inaccuracies.
> ```

Maybe add a comment to explain this rounding should not be tweaked unless care is taken it is the same as pg's
also add a reference to where the rounding occurs in pg (I sent you a link earlier)
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSYjc3EEjV5dwkAt7X:-LRSYjc3EEjV5dwkAt7Y:b61uxu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L37)):*
> ```Go
> 	ctx.WriteByte('(')
> 	for i, v := range d.D {
> 		if i > 0 {
> ```

the `comma` thing was better - more readable, also more efficient code (avoids a conditional; these are strictly worse than an extra assignment of a constant).
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 46 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZ2SRE6Z_Nlgl-R6e:-LRSZ2SRE6Z_Nlgl-R6f:b-ytug92) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L46)):*
> ```Go
> 		case *DCollatedString:
> 			pgwireFormatStringInTuple(ctx.Buffer, dv.Contents)
> 		case *DBytes:
> ```

Why is a separate case needed for DBytes? Please add a comment to explain/show why the `default` case below is not adequate.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 72 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZLKn9xyoocfUa3mW:-LRSZLKn9xyoocfUa3mX:bcw33ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L72)):*
> ```Go
> 			buf.WriteByte(byte(r))
> 		} else {
> 			buf.WriteRune(r)
> ```

Are you sure it's save to use `WriteRun` instead of `EncodeChar`? If there was an invalid utf-8 sequence, `r` will be invalid.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 94 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZYK85jHb_dWdDPN2:-LRSZYK85jHb_dWdDPN3:b-20ywe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L94)):*
> ```Go
> 	ctx.WriteByte('{')
> 	for i, v := range d.Array {
> 		if i > 0 {
> ```

ditto my previous comment about `comma`
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 104 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZa-SBJQCZ4fFMWrk:-LRSZa-SBJQCZ4fFMWrl:bta8ul9) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L104)):*
> ```Go
> 		case *DCollatedString:
> 			pgwireFormatStringInArray(ctx.Buffer, dv.Contents)
> 		case *DBytes:
> ```

ditto my previous comment about `DBytes`
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 133 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZhTE0t8tnf5YpLrQ:-LRSZhTE0t8tnf5YpLrR:babrp81) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L133)):*
> ```Go
> 
> func pgwireQuoteStringInArray(in string) bool {
> 	return in == "" || strings.ToLower(in) == "null" || arrayQuoteSet.in(in)
> ```

Do not call `ToLower` here. It forces an allocation, also if `in` is very long you're doing much too much work to determine what is needed. Better compare the 4 characters manually.
___
*[pkg/util/duration/duration.go, line 174 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8zIlVEcZqk0cPaT-e:-LRSVS2U8SMsCPD2fm3g:bilv5jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/util/duration/duration.go#L174)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I like this way of dealing with adding spaces.
</blockquote></details>

It would be even better to avoid the allocation of a closure and define `wrotePrev := func(wrote) bool { .... return wrote}`

then below write `wrote= wrotePrev(wrote)` every time.


<!-- Sent from Reviewable.io -->

madelynnblue committed:
pgwire: enable text encoding tests

madelynnblue committed:
pgwire: encoding tests/fixes for decimal and float nan/inf

Release note (bug fix): Correct binary decimal encoding for NaN.

madelynnblue committed:
pgwire: add encoding benchmarks

These use the same machinery as the encoding tests, so factor out some
common code.

Release note: None

madelynnblue reviewed (commented):
Added a skip to the initial test commit and removed it in the last commit.

Bram: we still accept dates/intervals in all the previous formats we did before. This change encodes them on the wire in the standard way.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-:-LRhn01z4YsxOOgHR2M3:b4vmx3v)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cli/testdata/dump/row, line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8zm9K1gDyEhe_jP0x:-LRhn_ZZ3MZaPd1M-F77:b2xfmhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/cli/testdata/dump/row#L93)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Was the extra empty line purposefully removed?
</blockquote></details>

I used the `-rewrite` flag to rewrite the file and it did this, so I think it's correct now.
___
*[pkg/cmd/cmp-protocol/main.go, line 68 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSWuy4CFNMvZln614F:-LRhnqnW8bRpDEhqf1iC:brt1ck0) ([raw file](https://github.com/cockroachdb/cockroach/blob/77ce7deda4c688ccc50f1a2ecee0f074baf06cd6/pkg/cmd/cmp-protocol/main.go#L68)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) Is there a way to include them, but have some kind of function which masks the last digits during the comparison to assert whether the test successful?


2) I'd like to see tests for DECIMAL and FLOAT which exercise the encodings for infinities and NaN. Are there any already? Perhaps somewhere else?
</blockquote></details>

1. The difficulty with float is that sometimes it's not just the last digit, since a .99999 could change to a 1.0. Also since this test just compares bytes and doesn't attempt to decode into datums, even doing that would be tricky.

2. Added a commit for float/decimal nan/inf.
___
*[pkg/sql/logictest/testdata/logic_test/time, line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR90ZkuBBrNUuB2sQZj:-LRiBCZtAFyB0iwN2ht8:bajp1de) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/sql/logictest/testdata/logic_test/time#L194)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

How do we handle `-00:00:00`?  Is it the same as `00:00:00`?  Please add tests for both of them.
</blockquote></details>

Tests added for both of these cases. (They are the same since -0 doesn't exist as an integer.)
___
*[pkg/sql/pgwire/types.go, line 197 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR91DtH02rvvZHhSFN0:-LRjcCLREHhPIBVZ5aIO:b-iiyyj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b052b7f6c93469861b9b2d9f63b07ccedb05b368/pkg/sql/pgwire/types.go#L197)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add a test for this?  Also, a comment here would be nice.
</blockquote></details>

There was a test for this already, and this commit fixed it.
___
*[pkg/sql/sem/tree/interval.go, line 562 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSUzvp2dB1vVy65Jot:-LRjekq28CU1J8moWAb1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/sql/sem/tree/interval.go#L562)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe add a comment to explain this rounding should not be tweaked unless care is taken it is the same as pg's
also add a reference to where the rounding occurs in pg (I sent you a link earlier)
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSYjc3EEjV5dwkAt7X:-LRjyayJ1CVqMghWt1Ip:b5apf9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L37)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

the `comma` thing was better - more readable, also more efficient code (avoids a conditional; these are strictly worse than an extra assignment of a constant).
</blockquote></details>

I'm genuinely surprised that it is more efficient, but a benchmark confirms. Reverted. Thanks for the insight.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 46 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZ2SRE6Z_Nlgl-R6e:-LRjfrMrE4Tjj2rXs6Am:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L46)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why is a separate case needed for DBytes? Please add a comment to explain/show why the `default` case below is not adequate.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 72 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZLKn9xyoocfUa3mW:-LRjz7ev0apEKEoxixNT:b8jfz1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L72)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Are you sure it's save to use `WriteRun` instead of `EncodeChar`? If there was an invalid utf-8 sequence, `r` will be invalid.
</blockquote></details>

This function is called in two ways: 1) with a string from a DString or CollatedString, 2) from a datum formatted into a string. My current understanding is that none of those can ever produce an invalid utf-8 sequence. If we can come up with some SQL that causes postgres to do the equivalent of EncodeChar then I would reconsider. But until then I would prefer to keep this code as is. That is, unless we can come up with a test that exhibits this behavior in pg, I think it makes more sense to keep it like this.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 94 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZYK85jHb_dWdDPN2:-LRjyl3H289dJI0tnhYD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L94)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto my previous comment about `comma`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 104 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZa-SBJQCZ4fFMWrk:-LRjfu0E-xfLQ5nZBXi3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L104)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto my previous comment about `DBytes`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 133 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSZhTE0t8tnf5YpLrQ:-LRjgxzH9EWDcHVxWEU3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L133)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Do not call `ToLower` here. It forces an allocation, also if `in` is very long you're doing much too much work to determine what is needed. Better compare the 4 characters manually.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/testdata/eval/overflow, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8xAB34aiZ9nu1upAq:-LRjhxvY05z1wwiyNMcM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026045098067c7d48d993ec293725ab4c79dacd/pkg/sql/sem/tree/testdata/eval/overflow#L1)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

no first line with an explanation
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/testdata/eval/row, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8xOiy6DUTxhwbVj9h:-LRjhO7v-P-2ZIg9Cq2F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026045098067c7d48d993ec293725ab4c79dacd/pkg/sql/sem/tree/testdata/eval/row#L1)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

most of these have a space after the first comment
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/testdata/eval/unary, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8xTFn15TXWW6J5sSb:-LRjhrzAC-A65MtlSlj9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e026045098067c7d48d993ec293725ab4c79dacd/pkg/sql/sem/tree/testdata/eval/unary#L1)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

No first line comment
</blockquote></details>

Done.
___
*[pkg/util/duration/duration.go, line 174 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR8zIlVEcZqk0cPaT-e:-LRjpyCtBdJsSEFmdmeV:bs4wn8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/util/duration/duration.go#L174)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It would be even better to avoid the allocation of a closure and define `wrotePrev := func(wrote) bool { .... return wrote}`

then below write `wrote= wrotePrev(wrote)` every time.
</blockquote></details>

Thanks for the suggestion. Using a func defined there that doesn't depend on arguments from the surrounding func performed 10% faster. Interestingly, if I just move that func definition to the package level, it gets 10% slower. Weird.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but now that you have a benchmark, please consider including a before/after benchmark result that reflects the changes from this entire PR. Thank you.

> Reviewed 33 of 37 files at r11, 29 of 36 files at r12, 1 of 3 files at r15, 26 of 30 files at r16, 3 of 3 files at r18, 1 of 1 files at r19.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-:-LRql4zq0ZsBq3li8PCm:b9sg158)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/array, line 0 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqjibi8y5mQrN6gSJl:-LRqjibj-Z560Y0_xgY0:beqxvct) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/logictest/testdata/logic_test/array#L0)):*
For some reason Reviewable thinks this file is "binary". Can you double check the git attributes?
___
*[pkg/sql/pgwire/types.go, line 245 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqkD9hEOddJXtDgMX6:-LRqkD9hEOddJXtDgMX7:bcbs0cv) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/pgwire/types.go#L245)):*
> ```Go
> 
> 	case *tree.DDecimal:
> 		if v.Form != apd.Finite {
> ```

Please file an issue and mention it in a TODO here.
___
*[pkg/sql/pgwire/types.go, line 257 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqkG1b3S7RaidCu5XJ:-LRqkG1b3S7RaidCu5XK:b63e0j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/pgwire/types.go#L257)):*
> ```Go
> 				// doesn't exist in postgres, it's unclear what to do. For now use the NaN
> 				// encoding and count it to see if anyone even needs this.
> 				telemetry.Count("pgwire.binary_decimal_infinity")
> ```

Add the issue number you'll create above to the counter name. We do this in other places already.
___
*[pkg/sql/sem/tree/pgwire_encode.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRSYjc3EEjV5dwkAt7X:-LRqk_H_92LaLgUYNFvC:bl2pu0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/100e2aeb2a4a8fe64f2ca39ef74c5d2395336074/pkg/sql/sem/tree/pgwire_encode.go#L37)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I'm genuinely surprised that it is more efficient, but a benchmark confirms. Reverted. Thanks for the insight.
</blockquote></details>

I learned from @vivekmenezes :)


<!-- Sent from Reviewable.io -->

knz commented:
Also I'll want to have a last look once you determine why there are still some tests failing.

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/array, line  at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqjibi8y5mQrN6gSJl:-LRtqw-4-OpuPj7CZdig:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/logictest/testdata/logic_test/array#L)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

For some reason Reviewable thinks this file is "binary". Can you double check the git attributes?
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/types.go, line 245 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqkD9hEOddJXtDgMX6:-LRtqxdq3wVJ6Hx6r_Bp:b-yr0903) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/pgwire/types.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please file an issue and mention it in a TODO here.
</blockquote></details>

See the first line of the TODO a few lines below. It lists the issue number.
___
*[pkg/sql/pgwire/types.go, line 257 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqkG1b3S7RaidCu5XJ:-LRtrZIBBrtqVmLxdyV7:bbdo9fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/pgwire/types.go#L257)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Add the issue number you'll create above to the counter name. We do this in other places already.
</blockquote></details>

I wasn't able to grep for any examples of this so I'm not sure of the standard. Can you link me one?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 93 of 93 files at r20, 3 of 3 files at r21, 37 of 37 files at r22, 36 of 36 files at r23, 2 of 2 files at r24, 1 of 1 files at r25, 3 of 3 files at r26, 31 of 31 files at r27, 1 of 1 files at r28, 3 of 3 files at r29, 1 of 1 files at r30.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/pgwire/types.go, line 257 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LRqkG1b3S7RaidCu5XJ:-LRvz0CdEJyWM0Fr-vg0:b-d0zisi) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c44ccd49d4228528aa0f432415506ca008fd9e/pkg/sql/pgwire/types.go#L257)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I wasn't able to grep for any examples of this so I'm not sure of the standard. Can you link me one?
</blockquote></details>

`pkg/server/updates_test.go`

```go
        expectedFeatureUsage := map[string]int32{
                "test.a": 1,
                "test.b": 2,
                "test.c": 3,

                "unimplemented.pg_catalog.pg_stat_wal_receiver":      10,
                "unimplemented.syntax.alter table rename constraint": 10,
                "unimplemented.syntax.interval with precision":       10,
                "unimplemented.#9148":                                10,
                "internalerror.":                                     10,
                "othererror.builtins.go":                             10,
                "othererror." +
                        pgerror.CodeDataExceptionError +
                        ".crdb_internal.set_vmodule()": 10,
                "errorcodes.blah":                                    10,
                "errorcodes." + pgerror.CodeDataExceptionError:       10,
                "errorcodes." + pgerror.CodeInternalError:            10,
                "errorcodes." + pgerror.CodeSyntaxError:              10,
                "errorcodes." + pgerror.CodeFeatureNotSupportedError: 10,
                "errorcodes." + pgerror.CodeDivisionByZeroError:      10,
        }
```

So you can use `pgwire.#NNNN.binary_decimal_infinity`


<!-- Sent from Reviewable.io -->

mjibson commented:
Added benchmarks to the array/tuple commit (since it's the only one with significant perf differences). There's overall speedup because (I think) of no longer calling into a unicode function per char. Ready for a final look.

knz reviewed (commented):
code LGTM still

1) regarding this:

> Added benchmarks to the array/tuple commit (since it's the only one with significant perf differences).

The part "it's the only one with signififcant perf differnces" -- how did you ascertain this? How can you explain me why you think this is true? Would you perhaps like to show me a copy of the benchmarks you used to form this opinion? If there were benchmarks, it wouldn't hurt including them in the PR description or commit message. If there were no benchmarks, you should still include the evidence that enables you to hold that opinion.

2) @bramgruneir please check whether your comments have been suitably addressed

3) for the newly added telemetry consider adding a test that the telemetry gets activated. See `TestErrorCounts` for an example.

> Reviewed 94 of 94 files at r31, 3 of 3 files at r32, 37 of 37 files at r33, 36 of 36 files at r34, 2 of 2 files at r35, 1 of 1 files at r36, 3 of 3 files at r37, 32 of 32 files at r38, 1 of 1 files at r39, 3 of 3 files at r40, 1 of 1 files at r41.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-:-LSEAai8A0urQm0OlokN:bp64dsh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just some new tests that I think were missing still, but I'm not sure.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-:-LSG3w9V4ZG3O4gu5tar:b-arq6il)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/time, line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32144#-LR90ZkuBBrNUuB2sQZj:-LSG3grPBnQ0DiX7JJ56:b-xw2ll8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c25ebfa3c782d99f97571f2750361d659727f1b7/pkg/sql/logictest/testdata/logic_test/time#L194)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Tests added for both of these cases. (They are the same since -0 doesn't exist as an integer.)
</blockquote></details>

I don't see the tests, but this is a huge diff, maybe I just missed it.


<!-- Sent from Reviewable.io -->

mjibson commented:
Added a telemetry test and benchmarks for intervals. The other encoding changes were bug fixes so I'm not going to benchmark them.

The interval benchmarks are ~30% slower. I think this is acceptable due to postgres parity.

knz commented:
Thanks for adding the benchmark. If you could add the before/after comparison at the top level PR description that would be swell.

mjibson commented:
bors r+
